Conditionalize warnings for IUO-to-Any coercion on Swift version.

rudkx opened the pull request:
These warnings are turning out to be pretty noisy for code that
declares IUOs (e.g. for @IBOutlets) and then passes them to
Objective-C APIs with parameters declared as _Nonnull id.

Since we bridge non-nil values successfully in most cases, and
previuosly written and correctly executing code is either not seeing
nil values passed in or are handling the nil (which is bridged as
NSNull), it seems like a nuisance to warn about these for existing
Swift versions.

We'll conditionalize the warning, and then users can deal with these
when moving to the new language version.

Fixes: rdar://problem/39886178

Commit:
Conditionalize warnings for IUO-to-Any coercion on Swift version.

These warnings are turning out to be pretty noisy for code that
declares IUOs (e.g. for @IBOutlets) and then passes them to
Objective-C APIs with parameters declared as _Nonnull id.

Since we bridge non-nil values successfully in most cases, and
previuosly written and correctly executing code is either not seeing
nil values passed in or are handling the nil (which is bridged as
NSNull), it seems like a nuisance to warn about these for existing
Swift versions.

We'll conditionalize the warning, and then users can deal with these
when moving to the new language version.

Fixes: rdar://problem/39886178

Commit:
Add a couple additional tests for Optional-to-Any warnings.

rudkx commented:
@swift-ci Please smoke test

jrose-apple commented on test/Sema/diag_unintended_optional_behavior.swift:
```diff
@@ -82,34 +82,33 @@ func warnNestedOptionalToOptionalAnyCoercion(_ a: Int?, _ b: Any??, _ c: Int???,
+  var a: Int!
+  var b: Any?!
+  func returningIUO() -> Int! { return a }
+  func returningAny() -> Any { return a }
```
For completeness can you test a subscript and an initializer too?

rudkx replied:
Good call, especially since I forgot to add subscript support!

I had made a note of doing so but then never came back and actually did it in the original commit.

jrose-apple reviewed (commented):
Code changes look good to me.

Commit:
Properly handle subscripts when avoiding IUO-to-Any coercion warnings.

Ensure that we grab the decl from the subscript expression so that we
can check whether it returns an IUO.

Also add tests for subscripts and failable inits.

rudkx commented:
@swift-ci Please smoke test

xedin reviewed (approved):
LGTM! Took me a bit do understand why there are `getDirectCallee` and `getCalledValue`.
